frontend/apps/reader/modules/readerpagemap.lua

```diff
@@ -0,0 +1,419 @@
+end
+
+function ReaderPageMap:getLastPageLabel(clean_label)
+    local label = self.ui.document:getPageMapLastPageLabel()
+    return clean_label and self:cleanPageLabel(label) or label
+end
+
+function ReaderPageMap:getXPointerPageLabel(xp, clean_label)
+    local label = self.ui.document:getPageMapXPointerPageLabel(xp)
+    return clean_label and self:cleanPageLabel(label) or label
+end
+
+function ReaderPageMap:getRenderedPageNumber(page_label, cleaned)
+    -- Only used from ReaderGoTo. As page_label is a string, no
+    -- way to use a binary search: do a full scan of the PageMap
+    -- here in Lua, even if it's not cheap.
+    local page_list = self.ui.document:getPageMap()
+    local rendered_page_number
+    for k, v in ipairs(page_list) do
+        local label = cleaned and self:cleanPageLabel(v.label) or v.label
+        if label == page_label then
+            return v.page
+        end
+    end
+end
+
+function ReaderPageMap:addToMainMenu(menu_items)
+    menu_items.page_map = {
+        text = _("Source pages"),
+        sub_item_table ={
+            {
+                text = _("Source info"),
+                enabled_func = function() return self.ui.document:getPageMapSource() ~= nil end,
+                callback = function()
+                    local text = T(_("Source (book hardcopy edition) of page numbers:\n\n%1"),
+                                    self.ui.document:getPageMapSource())
+                    local InfoMessage = require("ui/widget/infomessage")
+                    local infomsg = InfoMessage:new{
+                        text = text,
+                    }
+                    UIManager:show(infomsg)
+                end,
+                keep_menu_open = true,
+            },
+            {
+                text = _("Source pages list"),
+                callback = function()
+                    self:onShowPageList()
+                end,
+            },
+            {
+                text = _("Use source page numbers"),
+                checked_func = function() return self.use_page_labels end,
+                callback = function()
+                    self.use_page_labels = not self.use_page_labels
+                    self.ui.doc_settings:saveSetting("pagemap_use_page_labels", self.use_page_labels)
+                    -- Reset a few stuff that may use page labels
+                    self.ui.toc:resetToc()
+                    self.ui.view.footer:updateFooter()
+                    UIManager:setDirty(self.view.dialog, "partial")
+                end,
+                hold_callback = function(touchmenu_instance)
+                    local use_page_labels = G_reader_settings:isTrue("pagemap_use_page_labels")
+                    UIManager:show(MultiConfirmBox:new{
+                        text = use_page_labels and _("The default (★) for newly opened books that have a page map is to use these page labels instead of the renderer page numbers.\n\nWould you like to change it?")
```

Comment:
Not ecstatic about "renderer page numbers" but atm I can't think of a good alternative.

Reply:
Yeah, this one was hard.
I wanted to have the buttons "Renderer pages" | "Source pages (★)" but these got truncated...
So, we could end up with a sentence and Disable | Enable.